lxd/db/cluster/secrets.go

```diff
@@ -0,0 +1,270 @@
+}
+
+// newAuthSecret mints a new AuthSecret.
+func newAuthSecret() AuthSecret {
+	return AuthSecret{
+		Value:     newAuthSecretValue(),
+		CreatedAt: time.Now().UTC(),
+	}
+}
+
+// Validate checks that the secret is of the correct length and has not expired.
+func (s AuthSecret) Validate(expiry string) error {
+	err := s.Value.Validate()
+	if err != nil {
+		return err
+	}
+
+	expiresAt, err := shared.GetExpiry(s.CreatedAt, expiry)
+	if err != nil {
+		return fmt.Errorf("Failed to check auth secret expiry: %w", err)
+	}
+
+	if time.Now().UTC().After(expiresAt) {
+		return errors.New("Secret has expired")
+	}
+
+	return nil
+}
+
+// AuthSecrets is a concrete type for a slice of AuthSecret.
+type AuthSecrets []AuthSecret
+
+// Validate checks that there is at least one secret in the slice, and that the first (most recent) is valid and has not
+// expired.
+func (s AuthSecrets) Validate(expiry string) error {
+	if len(s) == 0 {
+		return errors.New("No secrets are defined")
+	}
+
+	return s[0].Validate(expiry)
+}
+
+// Rotate returns a new AuthSecrets, with a new AuthSecret prepended to the slice. The slice is always truncated to a
+// maximum of two elements. The new AuthSecret is written to the database, and any AuthSecret values older than the new
+// oldest in-memory value are deleted from the database.
+func (s AuthSecrets) Rotate(ctx context.Context, tx *sql.Tx) (AuthSecrets, error) {
+	// Get new secret.
+	newSecret := newAuthSecret()
+
+	// Initial in-memory rotation.
+	rotatedSecrets := append([]AuthSecret{newSecret}, s...)
+
+	// Truncation.
+	truncated := len(rotatedSecrets) > 2
+	if truncated {
+		rotatedSecrets = rotatedSecrets[:2]
+	}
+
+	// Add the new value to the database
+	err := createCoreAuthSecret(ctx, tx, newSecret)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to rotate secrets: %w", err)
+	}
+
+	// If the secrets were truncated, then there should be at most one old secret in the database. Delete it and assert
+	// that we deleted one row.
+	if truncated {
+		nDeleted, err := deleteCoreAuthSecretsOlderThanTimestamp(ctx, tx, rotatedSecrets[len(rotatedSecrets)-1].CreatedAt)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to delete out of date secrets: %w", err)
+		}
+
+		if nDeleted != 1 {
+			return nil, fmt.Errorf("Failed to delete expected number of out of date secrets: %w", err)
+		}
+	}
+
+	return rotatedSecrets, nil
+}
+
+// GetCoreAuthSecrets returns the server core auth secrets.
+func GetCoreAuthSecrets(ctx context.Context, tx *sql.Tx) (AuthSecrets, error) {
+	return getAuthSecrets(ctx, tx, entity.TypeServer, 0, SecretTypeCoreAuth)
+}
+
+// getAuthSecrets returns a slice of AuthSecrets. It scans the "secrets" table for rows of the given [entity.Type], entity
+// ID, and SecretType. Since we are scanning for AuthSecrets, the "value" column must be an AuthSecretValue.
+func getAuthSecrets(ctx context.Context, tx *sql.Tx, entityType entity.Type, entityID int, secretType SecretType) (AuthSecrets, error) {
```

Comment:
This seems like it should be named `getSecrets()` as it doesn't filter just by a specific secret type but rather accepts an argument.

Also the return type of `AuthSecrets` seems too specific and wouldn't we first have a struct type that represents a row in the secrets DB table and then we can convert that to a more specific representation if needed?

I'm a bit concerned we're going a bit type heavy TBH.

Reply:
Will remove this function and move the contents to GetAuthSecrets.